### PR TITLE
 FEC-11577 AdLayout Configuration

### PR DESCRIPTION
**Pull request by GouravSna**
### Description of the Changes

[Works with [Playkit-IMA](https://github.com/kaltura/playkit-android-ima)]

With Ad layout config you can create your own ad break timeline using your vast tags. Ad break can be set as pre, mid and post rolls and each ad break can contain a single vast tag or multiple tags, either as a pod, but also as a Waterfall.

### CODE REVIEW

File: tvplayer/src/main/java/com/kaltura/tvplayer/KalturaPlayer.java [hunk 1]
```diff
@@ -123,9 +126,11 @@
 
     private PrepareState prepareState = PrepareState.not_prepared;
     private PlayerTokenResolver tokenResolver = new PlayerTokenResolver();
+    private @Nullable PKAdvertisingController pkAdvertisingController;
     private PlayerInitOptions initOptions;
     private PlaylistController playlistController;
     private OfflineManager offlineManager;
+    private AdvertisingConfig advertisingConfig;
```
Comment (giladna) [inline tvplayer/src/main/java/com/kaltura/tvplayer/KalturaPlayer.java]: please put advertising next to each other

File: tvplayer/src/main/java/com/kaltura/tvplayer/KalturaPlayer.java [hunk 2]
```diff
@@ -543,11 +552,22 @@ public void onEvent(PlayerEvent event) {
                 pkPlayer.removeListener(this);
             }
         });
+
+        if (advertisingConfig != null) {
+            getPkAdvertisingController().playAdvertising();
```
Comment (giladna) [inline tvplayer/src/main/java/com/kaltura/tvplayer/KalturaPlayer.java]: what this code is doing?

File: tvplayer/src/main/java/com/kaltura/tvplayer/KalturaPlayer.java [hunk 3]
```diff
@@ -534,7 +539,11 @@ public void prepare() {
                 .setMediaEntry(mediaEntry)
                 .setStartPosition(startPosition);
 
+        pkPlayer.setAdvertising(advertisingConfig, getPkAdvertisingController());
+        getPkAdvertisingController().setPlayer(pkPlayer, messageBus, config);
+
```
Comment (giladna) [inline tvplayer/src/main/java/com/kaltura/tvplayer/KalturaPlayer.java]: @GouravSna please protect this code

File: tvplayer/src/main/java/com/kaltura/tvplayer/KalturaPlayer.java [hunk 4]
```diff
@@ -1308,6 +1332,74 @@ private void sendKavaPlayRequest(String entryId) {
         NetworkUtils.sendKavaAnalytics(context, partnerId, entryId, NetworkUtils.KAVA_EVENT_PLAY_REQUEST, sessionId);
     }
 
+    private PKAdvertisingController getPkAdvertisingController() {
+        if (pkAdvertisingController == null) {
```
Comment (giladna) [inline tvplayer/src/main/java/com/kaltura/tvplayer/KalturaPlayer.java]: should we create one if it is not needed

File: tvplayer/src/main/java/com/kaltura/tvplayer/KalturaPlayer.java [hunk 5]
```diff
@@ -1308,6 +1332,74 @@ private void sendKavaPlayRequest(String entryId) {
...
+
+        String imaPlugin = KnownPlugin.ima.name();
+        if (initOptions.pluginConfigs != null && initOptions.pluginConfigs.hasConfig(imaPlugin)) {
+            this.advertisingConfig = advertisingConfig;
+        } else {
+            log.w("IMAPlugin needs to be configured in order to use Advertising feature. \n " +
+                    "You can pass empty adtag url while configuring IMAPlugin");
+        }
+    }
+
```
Comment (giladna) [inline tvplayer/src/main/java/com/kaltura/tvplayer/KalturaPlayer.java]: lets make the public API short and call internal unified logic for both cases

File: tvplayer/src/main/java/com/kaltura/tvplayer/KalturaPlayer.java [hunk 6]
```diff
@@ -1308,6 +1332,54 @@ private void sendKavaPlayRequest(String entryId) {
...
+        if (initOptions.pluginConfigs != null && initOptions.pluginConfigs.hasConfig(imaPlugin)) {
+            if (advertising instanceof String) {
+                try {
+                    advertisingConf = new Gson().fromJson((String) advertising, AdvertisingConfig.class);
+                    if (advertisingConf != null) {
+                        this.advertisingConfig = advertisingConf;
+                    } else {
+                        log.e("Malformed AdvertisingConfig Json");
+                    }
+                } catch (JsonSyntaxException e) {
```
Comment (GouravSna) [inline tvplayer/src/main/java/com/kaltura/tvplayer/KalturaPlayer.java]: Try putting some junk json and see the exception

File: tvplayer/src/main/java/com/kaltura/tvplayer/KalturaPlayer.java (hunk 2 above)
Comment (GouravSna) [reply tvplayer/src/main/java/com/kaltura/tvplayer/KalturaPlayer.java]: Removed this code completely.

File: tvplayer/src/main/java/com/kaltura/tvplayer/KalturaPlayer.java (hunk 3 above)
Comment (GouravSna) [reply tvplayer/src/main/java/com/kaltura/tvplayer/KalturaPlayer.java]: Done

File: tvplayer/src/main/java/com/kaltura/tvplayer/KalturaPlayer.java (hunk 4 above)
Comment (GouravSna) [reply tvplayer/src/main/java/com/kaltura/tvplayer/KalturaPlayer.java]: Yes, Done

File: tvplayer/src/main/java/com/kaltura/tvplayer/KalturaPlayer.java (hunk 5 above)
Comment (GouravSna) [reply tvplayer/src/main/java/com/kaltura/tvplayer/KalturaPlayer.java]: Done. Moved it in a single method receiving object.

**Review by giladna — APPROVED** [review]

**Pull request merged**